entmax/root_finding.py

```diff
@@ -133,6 +134,129 @@ def backward(cls, ctx, dY):
+        tau_lo = max_val - cls._gp(1, alpha)  # 1
+        tau_hi = max_val - cls._gp(1 / d, alpha)  # (1/d)**(alpha-1)
+
+        f_lo = (cls._p(X - tau_lo, alpha) ** alpha).sum(dim) - 1
+        dm = tau_hi - tau_lo
+
+        for it in range(n_iter):
+            dm /= 2
+            tau_m = tau_lo + dm
+            p_m = cls._p(X - tau_m, alpha)  # [X - tau]_+ ** (1/(alpha-1))
+            f_m = (p_m ** alpha).sum(dim) - 1
+            mask = (f_m >= 0).unsqueeze(dim)
+            tau_lo = torch.where(mask, tau_m, tau_lo)
+
+        p_m /= p_m.sum(dim=dim).unsqueeze(dim=dim)
+
+        ctx.save_for_backward(p_m)
+
+        return p_m
+
+    @classmethod
+    def backward(cls, ctx, dY):
+        Y, = ctx.saved_tensors
+
+        a = torch.where(Y > 0, Y, Y.new_zeros(1))
+        b = torch.where(Y > 0, Y ** (2 - ctx.alpha), Y.new_zeros(1))
+        
+        dX = dY * b
+        q = dX.sum(ctx.dim).unsqueeze(ctx.dim)
+        dX -= q * a
+        q = (dY * a).sum(ctx.dim).unsqueeze(ctx.dim)
+        dX -= q * (b - b.sum(ctx.dim).unsqueeze(ctx.dim) * a)
+        dX *= ((a ** ctx.alpha).sum(ctx.dim).unsqueeze(ctx.dim) **
+               ((ctx.alpha - 1) / ctx.alpha))
+        dX /= (ctx.alpha - 1)
+
+        return dX, None, None, None, None
+
+
+class BudgetBisectFunction(Function):
+    @classmethod
+    def forward(cls, ctx, X, budget=2, dim=-1, n_iter=50):
+
+        if not isinstance(budget, torch.Tensor):
+            budget = torch.tensor(budget, dtype=X.dtype, device=X.device)
+
+        budget_shape = list(X.shape)
+        budget_shape[dim] = 1
+        budget = budget.expand(*budget_shape)
+
+        ctx.budget = budget
+        ctx.dim = dim
+        d = X.shape[dim]
+
+        max_val, _ = X.max(dim=dim, keepdim=True)
+        min_val, _ = X.min(dim=dim, keepdim=True)
+
+        # If tau = max_val - B/N, X-tau <= B/N, hence sum(max(0, X-tau)) <= B.
+        # If tau = min_val - B/N, X-tau >= B/N, hence sum(min(1, X-tau)) >= B.
+        tau_lo = min_val - budget / X.shape[dim]
+        tau_hi = max_val - budget / X.shape[dim]
+
+        f_lo = (torch.clamp(X - tau_lo, min=0, max=1).sum(dim).unsqueeze(dim)
+                - budget)
+        dm = tau_hi - tau_lo
+
+        for it in range(n_iter):
+            dm /= 2
+            tau_m = tau_lo + dm
+            p_m = torch.clamp(X - tau_m, min=0, max=1)
+            f_m = p_m.sum(dim).unsqueeze(dim) - budget
+            mask = (f_m >= 0)
+            tau_lo = torch.where(mask, tau_m, tau_lo)
+
+        ctx.save_for_backward(p_m)
+
+        return p_m
+
+    @classmethod
+    def backward(cls, ctx, dY):
+        Y, = ctx.saved_tensors
+
+        # tau = (size_one  + sum(score[active]) - budget) / size_active
+        # dtau / dscore = 1 / size_active
+        # J = Diag(s) - ss' / size_active, s indicator vector of active indices.
+
+        active = (Y > 0) & (Y < 1)
+        active_size = active.sum(ctx.dim)
+        v_hat = ((dY * active).sum(ctx.dim) /
```

Comment:
This is the problem line. @andre-martins @vene everyone okay with increasing the minimum version to 1.2 so this is fine?